src/mds/Server.cc

```diff
@@ -2237,20 +2252,39 @@ CDentry* Server::prepare_null_dentry(MDRequestRef& mdr, CDir *dir, const string&
 
 CDentry* Server::prepare_stray_dentry(MDRequestRef& mdr, CInode *in)
 {
+  string straydname;
+  in->name_stray_dentry(straydname);
   CDentry *straydn = mdr->straydn;
   if (straydn) {
-    string name;
-    in->name_stray_dentry(name);
-    if (straydn->get_name() == name)
+    if (straydn->get_name() == straydname)
       return straydn;
 
     assert(!mdr->done_locking);
     mdr->unpin(straydn);
   }
 
-  straydn = mdcache->get_or_create_stray_dentry(in);
+  CDir *straydir = mdcache->get_stray_dir(in);
+
+  if (!check_fragment_space(mdr, straydir))
+    return NULL;
+
+  straydn = straydir->lookup(straydname);
+  if (!straydn) {
+    straydn = straydir->add_null_dentry(straydname);
+    straydn->mark_new();
+  } else {
+    assert(straydn->get_projected_linkage()->is_null());
+  }
+
+  // Notify even if a null dentry already existed, because
+  // StrayManager counts the number of stray inodes, not the
+  // number of dentries in the directory.
+  mdcache->notify_stray_created();
+
+  straydn->state_set(CDentry::STATE_STRAY);
```

Comment:
Please don't open-code the "create" bit of "get_or_create_stray_dentry". It belongs more in MDCache, and if you just put the get_stray_dir(CInode*) and create_stray_dentry(CDir* stray, const string& name) next to each other the diff will be smaller as well. ;)